p4-fusion/lfs/communication/s3comm.cc

```diff
@@ -0,0 +1,54 @@
+#include "s3comm.h"
+#include "lfs/lfs_client.h"
+#include <aws/core/Aws.h>
+#include <aws/core/Region.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/PutObjectRequest.h>
+#include <aws/core/auth/AWSCredentials.h>
+#include <aws/s3/S3ClientConfiguration.h>
+#include <aws/core/utils/stream/PreallocatedStreamBuf.h>
+
+S3Comm::S3Comm(const std::string& serverURL, const std::string& bucket, const std::string& repository, const std::string& username, const std::string& password)
+	: m_ServerURL(serverURL)
+	, m_Bucket(bucket)
+	, m_Repository(repository)
+	, m_Username(username)
+	, m_Password(password)
+{
+}
+
+Communicator::UploadResult S3Comm::UploadFile(const std::vector<char>& fileContents) const
+{
+	std::string oid = LFSClient::CalcOID(fileContents);
+
+	Aws::Auth::AWSCredentials credentials(m_Username, m_Password);
+
+	bool isHttps = m_ServerURL.find("https://") == 0;
+
+    Aws::S3::S3ClientConfiguration config;
+    config.endpointOverride = m_ServerURL;
+    config.scheme = isHttps ? Aws::Http::Scheme::HTTPS : Aws::Http::Scheme::HTTP;
+    config.verifySSL = isHttps;
+    config.region = Aws::Region::US_EAST_1; // MinIO ignores region, but AWS SDK requires it
+	config.useVirtualAddressing = false;
+
+	Aws::S3::S3Client s3_client(credentials, config,
+        Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Never, false);
+
+	Aws::S3::Model::PutObjectRequest object_request;
```

Comment:
A file with the very same content can appear in multiple changelists. If this happens, we will upload the same file more than once. This provides an optimization opportunity: before uploading a file, we can query whether it was already uploaded or not. Interestingly, LFS' basic transfer/batch API supports this out-of-the box, and `p4-fusion` already makes use of this.

I've made some measurements with converting the BIMx codebase to see whether this would be worth doing, or not. The results are quite convincing:
* conversion time: 27m vs. 31m (13% improvement)
* upload bandwidth saved: 1.97 GiB (only logged the savings, so I don't know the baseline, sorry!)

The code to achieve this is quite simple:

```suggestion
	// Check if object already exists
	Aws::S3::Model::HeadObjectRequest headRequest;
	head_request.SetBucket(m_Bucket);
	head_request.SetKey(objectKey);

	auto headResult = s3_client.HeadObject(headRequest);
	if (headResult.IsSuccess()) {
		return UploadResult::AlreadyExists;
	}
```

Reply:
My suggestion is to add this piece of code, and not to remove this line, messed up the "Add code suggestion" thingy.

Reply:
I've fixed this in b4fe3c24ad518656b80c693a6279ac1d0232beff.